File: Java/laborator3JavaGUI/src/main/java/repo/ParticipantCursaRepository.java
package repo;

import javafx.util.Pair;
import domain.ParticipantCursa;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantCursaRepository implements IParticipantCursaRepository {
    private static final Logger logger = LogManager.getLogger(ParticipantCursaRepository.class);
    private JdbcUtils jdbcUtils;

    public ParticipantCursaRepository(Properties properties) {
        logger.info("Initializing ParticipantCursaRepository with properties {}", properties);
        this.jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public void save(ParticipantCursa entity) {
        ParticipantCursa participantCursa = findOne(new Pair<>(entity.getId().getKey(), entity.getId().getValue()));
        if(participantCursa != null)
            throw new RepositoryException("Entity is already saved!");

        logger.info("Saving entity {}", entity);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into ParticipantCursa values (?,?)")){
            preparedStatement.setString(1, entity.getIdParticipant());
            preparedStatement.setString(2, entity.getIdCursa());
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(Pair<String, String> pair, ParticipantCursa entity) {
        logger.traceEntry("Updating entity with id {}", pair);
        logger.traceExit();
    }

    @Override
    public void delete(Pair<String, String> pair) {
        logger.traceEntry("Deleting entity with id {}", pair);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from ParticipantCursa where idParticipant=? and idCursa=?")){
            preparedStatement.setString(1, pair.getKey());
            preparedStatement.setString(2, pair.getValue());
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit();
    }

    @Override
    public ParticipantCursa findOne(Pair<String, String> pair) {
        logger.traceEntry("Finding entity with id {}", pair);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from ParticipantCursa where idParticipant=? and idCursa=?")){
            preparedStatement.setString(1, pair.getKey());
            preparedStatement.setString(2, pair.getValue());
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    String idParticipant = resultSet.getString("idParticipant");
                    String idCursa = resultSet.getString("idCursa");
                    ParticipantCursa participantCursa = new ParticipantCursa(idParticipant, idCursa);
                    participantCursa.setId(new Pair<>(idParticipant, idCursa));
                    logger.traceExit(participantCursa);
                    return participantCursa;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit("No entity with id {} found", pair);
        return null;
    }

    @Override
    public Iterable<ParticipantCursa> findAll() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        List<ParticipantCursa> participantCursaList = new ArrayList<>();

        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from ParticipantCursa")){
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                while (resultSet.next()){
                    String idParticipant = resultSet.getString("idParticipant");
                    String idCursa = resultSet.getString("idCursa");
                    ParticipantCursa participantCursa = new ParticipantCursa(idParticipant, idCursa);
                    participantCursa.setId(new Pair<>(idParticipant, idCursa));
                    participantCursaList.add(participantCursa);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit(participantCursaList);
        return participantCursaList;
    }

    @Override
    public int size() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("select count(*) as [size] from ParticipantCursa")){
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    logger.traceExit(resultSet.getInt("size"));
                    return resultSet.getInt("size");
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        return 0;
    }
}

File: C#/Aplicatii client-server/labClientServerCSharpProtoBuf/ClientServer/Persistence/ParticipantCursaRepository.cs
using lab3.domain;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using log4net;
using System.Data;

namespace lab3.repo
{
    public class ParticipantCursaRepository : ICrudRepository<Tuple<string, string>, ParticipantCursa>
    {
        private static readonly ILog log = LogManager.GetLogger("ParticipantCursaRepository");
        public ParticipantCursaRepository() { log.Info("Initializing ParticipantCursaRepository"); }
        public void delete(Tuple<string, string> id)
        {
            log.InfoFormat("Deleting entity with id [{0};{1}]", id.Item1, id.Item2);
            IDbConnection connection = DBUtils.getConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "delete from ParticipantCursa where idParticipant=@idP and idCursa=@idC";

                var paramIdP = command.CreateParameter();
                paramIdP.ParameterName = "@idP";
                paramIdP.Value = id.Item1;
                command.Parameters.Add(paramIdP);

                var paramIdC = command.CreateParameter();
                paramIdC.ParameterName = "@idC";
                paramIdC.Value = id.Item2;
                command.Parameters.Add(paramIdC);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No entity deleted!");
            }
        }

        public IEnumerable<ParticipantCursa> findAll()
        {
            IDbConnection connection = DBUtils.getConnection();
            IList<ParticipantCursa> participantCursaList = new List<ParticipantCursa>();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select idParticipant, idCursa from ParticipantCursa";

                using (var result = command.ExecuteReader())
                {
                    while (result.Read())
                    {
                        int idP = result.GetInt32(0);
                        int idC = result.GetInt32(1);
                        ParticipantCursa participantCursa = new ParticipantCursa(idP.ToString(), idC.ToString())
                        {
                            Id = new Tuple<string, string>(idP.ToString(), idC.ToString())
                        };
                        participantCursaList.Add(participantCursa);
                    }
                }
            }

            return participantCursaList;
        }

        public ParticipantCursa findOne(Tuple<string, string> id)
        {
            log.InfoFormat("Finding entity with id [{0};{1}]", id.Item1, id.Item2);
            IDbConnection connection = DBUtils.getConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText =
                    "select idParticipant, idCursa from ParticipantCursa where idParticipant=@idP and idCursa=@idC";
                IDbDataParameter paramIdP = command.CreateParameter();
                paramIdP.ParameterName = "@idP";
                paramIdP.Value = id.Item1;
                command.Parameters.Add(paramIdP);

                IDbDataParameter paramIdC = command.CreateParameter();
                paramIdC.ParameterName = "@idC";
                paramIdC.Value = id.Item2;
                command.Parameters.Add(paramIdC);

                using (var result = command.ExecuteReader())
                {
                    if (result.Read())
                    {
                        int idP = result.GetInt32(0);
                        int idC = result.GetInt32(1);
                        ParticipantCursa participantCursa = new ParticipantCursa(idP.ToString(), idC.ToString())
                        {
                            Id = new Tuple<string, string>(idP.ToString(), idC.ToString())
                        };
                        log.InfoFormat("Exiting with entity found {0}", participantCursa);
                        return participantCursa;
                    }
                }
            }

            log.InfoFormat("No entity with id [{0};{1}] found", id);
            return null;
        }

        public void save(ParticipantCursa entity)
        {
            if(findOne(entity.Id) != null)
            {
                throw new Exception("Entity already saved!");
            }

            log.InfoFormat("Saving entity {0}", entity);
            IDbConnection connection = DBUtils.getConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "insert into ParticipantCursa values (@idParticipant, @idCursa)";

                var paramIdP = command.CreateParameter();
                paramIdP.ParameterName = "@idParticipant";
                paramIdP.Value = entity.Id.Item1;
                command.Parameters.Add(paramIdP);

                var paramIdC = command.CreateParameter();
                paramIdC.ParameterName = "@idCursa";
                paramIdC.Value = entity.Id.Item2;
                command.Parameters.Add(paramIdC);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No entity added!");
            }
        }
    }
}
